internal/dag/cache.go

```diff
@@ -123,3 +123,69 @@ func (kc *KubernetesCache) remove(obj interface{}) {
 		// not interesting
 	}
 }
+
+// SecretReferenced looks at all Ingress & IngressRoute resources and determines
+// if a secret is referenced in either of those resources
+func (kc *KubernetesCache) SecretReferenced(secret *v1.Secret) bool {
+
+	for _, ingress := range kc.ingresses {
+		// Check if the Secret is referenced to an ingress resource
+		for _, tls := range ingress.Spec.TLS {
+			if tls.SecretName == secret.GetName() && ingress.GetNamespace() == secret.GetNamespace() {
+				return true
+			}
+		}
+	}
+	for _, ingress := range kc.ingressroutes {
+		// If VirtualHost or TLS is not defined then continue
+		if ingress.Spec.VirtualHost == nil || ingress.Spec.VirtualHost.TLS == nil {
+			continue
+		}
+		if ingress.Spec.VirtualHost.TLS.SecretName == secret.GetName() &&
+			ingress.GetNamespace() == secret.GetNamespace() {
+			return true
+		}
+	}
+	for _, delcrt := range kc.delegations {
+		// If the secret is referenced in a TLS delegation, then it's in use
+		for _, del := range delcrt.Spec.Delegations {
+			if del.SecretName == secret.GetName() && delcrt.GetNamespace() == secret.GetNamespace() {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// ServiceReferenced looks at all Ingress & IngressRoute resources and determines
+// if a service is referenced in either of those resources
+func (kc *KubernetesCache) ServiceReferenced(service *v1.Service) bool {
+	for _, ingress := range kc.ingresses {
+		if ingress.Spec.Backend != nil {
+			if ingress.Spec.Backend.ServiceName == service.GetName() && ingress.GetNamespace() == service.GetNamespace() {
+				return true
+			}
+		}
+		// Check if the Service is referenced to an ingress resource
+		for _, rule := range ingress.Spec.Rules {
+			if rule.IngressRuleValue.HTTP == nil {
+				continue
+			}
+			for _, path := range rule.IngressRuleValue.HTTP.Paths {
+				if path.Backend.ServiceName == service.GetName() && ingress.GetNamespace() == service.GetNamespace() {
+					return true
+				}
+			}
+		}
+	}
+	for _, ingress := range kc.ingressroutes {
+		for _, route := range ingress.Spec.Routes {
+			for _, routeSvc := range route.Services {
+				if routeSvc.Name == service.GetName() && ingress.GetNamespace() == service.GetNamespace() {
```

Comment:
I'm not sure this is right. This will say that any service is referenced if it is present in a ingressroute even if the ingressroute is not referenced from a root.